wagtail/core/collectors.py

```diff
@@ -0,0 +1,379 @@
+from wagtail.core.rich_text import features, RichText
+from wagtail.core.rich_text.rewriters import FIND_A_TAG, FIND_EMBED_TAG, extract_attrs
+from wagtail.utils.pagination import paginate
+
+
+def get_obj_base_key(obj):
+    if isinstance(obj, Model):
+        base_model = ([obj._meta.model] + obj._meta.get_parent_list())[-1]
+        return base_model._meta.label, obj.pk
+    return obj
+
+
+def find_objects_in_rich_text(rich_text: str):
+    for regex, registry, name_attr in (
+            (FIND_A_TAG, features.get_link_types(), 'linktype'),
+            (FIND_EMBED_TAG, features.get_embed_types(), 'embedtype')):
+        for attr_string in regex.findall(rich_text):
+            attrs = extract_attrs(attr_string)
+            if name_attr not in attrs:
+                continue
+            handler = registry[attrs[name_attr]]
+            instance = handler.get_instance(attrs)
+            if instance is not None:
+                yield instance
+
+
+class ModelRichTextCollector:
+    def __init__(self, model, using=DEFAULT_DB_ALIAS):
+        self.model = model
+        self.using = using
+        self.fields = [f for f in self.model._meta.fields
+                       if isinstance(f, RichTextField)]
+
+    @staticmethod
+    def get_handlers(searched_objects):
+        if searched_objects:
+            searched_models = set()
+            for obj in searched_objects:
+                searched_models.add(obj._meta.model)
+                searched_models.update(obj._meta.get_parent_list())
+
+        for handler in chain(features.get_link_types().values(),
+                             features.get_embed_types().values()):
+            model = handler.get_model()
+            if searched_objects:
+                if model in searched_models:
+                    yield model, handler
+            else:
+                yield model, handler
+
+    @classmethod
+    def get_pattern_for_objects(cls, searched_objects):
+        handlers = dict(cls.get_handlers(searched_objects))
+        if not handlers:
+            return
+
+        link_types = [re.escape(h.link_type) for h in handlers.values()]
+        type_pat = r'(link|embed)type="%s"' % (
+            link_types[0] if len(link_types) == 1
+            else r'(%s)' % r'|'.join(link_types))
+        params = {'type': type_pat}
+        if searched_objects:
+            pattern = r'<(a|embed)( %(type)s %(val)s| %(val)s %(type)s)[^>]*>'
+            values = []
+            for obj in searched_objects:
+                for model, handler in handlers.items():
+                    if isinstance(obj, model):
+                        k, v = handler.get_id_pair_from_instance(obj)
+                        values.append('%s="%s"' % (k, re.escape(str(v))))
+            params['val'] = (values[0] if len(values) == 1
+                             else r'(%s)' % r'|'.join(values))
+        else:
+            pattern = r'<(a|embed) %(type)s[^>]*>'
```

Comment:
From what I can see, this case exists to serve a "hidden feature" of `find_objects` where passing an empty object list returns all objects of any type referenced within rich text. What's the motivation for this feature? It seems out of place here: it's listing objects rather than finding them; it doesn't match the logical expected behaviour of find_object on an empty list; and it requires quite a bit of special-case code to make it work (and, in particular, conflicts with my earlier assertion that link/embed handlers don't have to be associated with a model)...

Reply:
Yes, this is indeed an unused feature for now. At first, I started with doing that wildcard search only if `searched_objects is None`.
If I remember well, I changed my mind to make a cleaner API using `.find_objects(*searched_objects)` instead of `.find_objects(searched_objects=None)`. I’m sure I thought “if you search for one or more objects, it searches them, otherwise it searches for every object”. It must have been a brain fart, as I clearly forgot that sometimes we just pass a tuple or list directly without checking whether an item exists or not.

The good news is that it cannot have any impact on the views added, but that’s definitely worth changing. I think that we should just modify line 79 with `if searched_objects is None` and reverse the condition, then modify `find_objects`, for example by adding a `find_all` boolean that could only be `True` if there’s no `searched_objects`. Or add a `_find_objects(searched_objects)` method with all the logic from `find_objects`, and create two methods `find_objects(*searched_objects)` and `find_all_objects()` (that last one passing `None` to `_find_objects`).

What do you think?

Reply:
Having two separate methods for `find_objects` and `find_all_objects` definitely makes more sense to me - they're functionally very different things. I think `get_pattern_for_objects` and `get_handlers` can also be split up in the same way, to simplify the logic.